### PR TITLE
invoke credentials builder from model initializer

### LINKED ISSUE #213: Model initializer does not work with S3
/kind bug

**What steps did you take and what happened:**
The model download is now moved to `init container`, however the S3 secret envs are set in the `user-container`. 

kubectl logs mnist-default-g7px9-deployment-5d6d7f897c-t5rzk -c model-initializer 
INFO:root:Initializing, args: src_uri [s3://platform-test/mnist/export] dest_path[ [/mnt/models]
INFO:root:Copying contents of s3://platform-test/mnist/export to local
Traceback (most recent call last):
  File "/model-initializer/scripts/initializer-entrypoint", line 14, in <module>
    kfserving.Storage.download(src_uri, dest_path)
  File "/kfserving/kfserving/storage.py", line 41, in download
    Storage._download_s3(uri, out_dir)
  File "/kfserving/kfserving/storage.py", line 52, in _download_s3
    client = Storage._create_minio_client()
  File "/kfserving/kfserving/storage.py", line 105, in _create_minio_client
    secure=True)
  File "/usr/local/lib/python3.7/site-packages/minio/api.py", line 149, in __init__
    is_valid_endpoint(endpoint)
  File "/usr/local/lib/python3.7/site-packages/minio/helpers.py", line 301, in is_valid_endpoint
    if hostname[-1] == '.':
IndexError: string index out of range

**What did you expect to happen:**
Secret env and mount should be moved to init container.

### PR DESCRIPTION
**Pull request by rakelkar**
fixes #213

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/215)

### CODE REVIEW

File: pkg/webhook/admission/deployment/mutator.go [hunk 1]
```diff
@@ -42,7 +50,14 @@ func (mutator *Mutator) Handle(ctx context.Context, req types.Request) types.Res
 		return admission.ErrorResponse(http.StatusBadRequest, err)
 	}
 
-	if err := mutate(deployment); err != nil {
+	configMap := &v1.ConfigMap{}
+	err := mutator.Client.Get(context.TODO(), k8types.NamespacedName{Name: constants.KFServiceConfigMapName, Namespace: constants.KFServingNamespace}, configMap)
+	if err != nil {
+		log.Error(err, "Failed to find config map", "name", constants.KFServiceConfigMapName)
+		return admission.ErrorResponse(http.StatusBadRequest, err)
```
Comment (yuzisun) [inline pkg/webhook/admission/deployment/mutator.go]: this sounds like `http.StatusInternalServerError`

File: pkg/constants/constants.go [hunk 2]
```diff
@@ -45,8 +45,9 @@ var (
 
 // KFService Internal Annotations
 var (
-	KFServiceInternalAnnotationsPrefix             = "internal." + KFServingAPIGroupName
-	ModelInitializerSourceUriInternalAnnotationKey = KFServiceInternalAnnotationsPrefix + "/model-initializer-sourceuri"
+	KFServiceInternalAnnotationsPrefix                      = "internal." + KFServingAPIGroupName
+	ModelInitializerSourceUriInternalAnnotationKey          = KFServiceInternalAnnotationsPrefix + "/model-initializer-sourceuri"
+	ModelInitializerServiceAccountNameInternalAnnotationKey = KFServiceInternalAnnotationsPrefix + "/model-initializer-sa-name"
```
Comment (yuzisun) [inline pkg/constants/constants.go]: service account name is passed down on deployment `deployment.Spec.Template.Spec.ServiceAccountName`

File: pkg/webhook/admission/deployment/model_initializer_injector.go [hunk 3]
```diff
@@ -136,6 +141,20 @@ func InjectModelInitializer(deployment *appsv1.Deployment) error {
 	// Add volumes to the PodSpec
 	podSpec.Volumes = append(podSpec.Volumes, podVolumes...)
 
+	serviceAccountName, ok := annotations[constants.ModelInitializerServiceAccountNameInternalAnnotationKey]
+	if !ok {
```
Comment (ellis-bigelow) [inline pkg/webhook/admission/deployment/model_initializer_injector.go]: This is already "" right? Or even if it's nil, that's ok I think

File: pkg/webhook/admission/deployment/model_initializer_injector.go [hunk 4]
```diff
@@ -114,7 +119,7 @@ func InjectModelInitializer(deployment *appsv1.Deployment) error {
 	modelInitializerMounts = append(modelInitializerMounts, sharedVolumeWriteMount)
 
 	// Add an init container to run provisioning logic to the PodSpec
-	initContianer := v1.Container{
+	initContianer := &v1.Container{
```
Comment (ellis-bigelow) [inline pkg/webhook/admission/deployment/model_initializer_injector.go]: initContianer -> initContainer

**Comment by rakelkar** [comment]
removed sa logic - as noted in comments, its already there!

**Comment by rakelkar** [comment]
@ellis-bigelow @yuzisun  this is ready to merge. have added tests to verify the init container has the right creds

File: pkg/controller/kfservice/resources/credentials/service_account_credentials.go [hunk 5]
```diff
@@ -61,7 +60,7 @@ func NewCredentialBulder(client client.Client, config *v1.ConfigMap) *Credential
 }
 
 func (c *CredentialBuilder) CreateSecretVolumeAndEnv(namespace string, serviceAccountName string,
-	configuration *knservingv1alpha1.Configuration) error {
+	container *v1.Container, volumes *[]v1.Volume) error {
```
Comment (ellis-bigelow) [inline pkg/controller/kfservice/resources/credentials/service_account_credentials.go]: +1

Comment (ellis-bigelow) [reply pkg/controller/kfservice/resources/credentials/service_account_credentials.go]: Actually, better to pass in the template spec? It's ducktyped so it's generic to k8s v1 right?

Comment (rakelkar) [reply pkg/controller/kfservice/resources/credentials/service_account_credentials.go]: Not sure how to do that - can you do that in a later change?

**Comment by yuzisun** [comment]
cool! one open question left is that wether we can use custom container with model initializer, so that credential builder can be removed from user container.

/lgtm

**Comment by rakelkar** [comment]
I wonder if custom container should morph into "custom predictor" i.e. still support stuff like ModelUri or have some kind of env var contract so that we can find and replace the ModelUri

**Pull request merged**